File: 3455.shortest-matching-substring.py
#
# @lc app=leetcode id=3455 lang=python3
#
# [3455] Shortest Matching Substring
#

# @lc code=start
import bisect

class Solution:
    def shortestMatchingSubstring(self, s: str, p: str) -> int:
        def find_matches(text: str, pat: str) -> list[int]:
            n, m = len(text), len(pat)
            if m == 0 or m > n:
                return []
            # KMP prefix function
            pi = [0] * m
            j = 0
            for i in range(1, m):
                while j > 0 and pat[i] != pat[j]:
                    j = pi[j - 1]
                if pat[i] == pat[j]:
                    j += 1
                pi[i] = j
            # Search
            matches = []
            j = 0
            for i in range(n):
                while j > 0 and text[i] != pat[j]:
                    j = pi[j - 1]
                if text[i] == pat[j]:
                    j += 1
                if j == m:
                    matches.append(i - m + 1)
                    j = pi[j - 1]
            return matches

        star1 = p.find('*')
        star2 = p.find('*', star1 + 1)
        A = p[:star1]
        B = p[star1 + 1:star2]
        C = p[star2 + 1:]
        lenA, lenB, lenC = len(A), len(B), len(C)
        n = len(s)
        if lenA == 0 and lenB == 0 and lenC == 0:
            return 0

        startsA = find_matches(s, A) if lenA > 0 else []
        startsB = find_matches(s, B)
        startsC = find_matches(s, C) if lenC > 0 else []

        min_len = float('inf')

        if lenB > 0:
            for kb in startsB:
                target_jc = kb + lenB
                if lenC == 0:
                    jc_next = target_jc
                else:
                    idx = bisect.bisect_left(startsC, target_jc)
                    if idx == len(startsC):
                        continue
                    jc_next = startsC[idx]
                j = jc_next + lenC - 1
                target_ia = kb - lenA
                if target_ia < 0:
                    continue
                if lenA == 0:
                    i_start = target_ia
                else:
                    idx = bisect.bisect_right(startsA, target_ia) - 1
                    if idx < 0:
                        continue
                    i_start = startsA[idx]
                if i_start > j:
                    continue
                length = j - i_start + 1
                if length < min_len:
                    min_len = length
        else:
            # lenB == 0
            startsA = find_matches(s, A) if lenA > 0 else []
            startsC = find_matches(s, C) if lenC > 0 else []
            if lenC == 0:
                if lenA == 0:
                    min_len = 0
                elif startsA:
                    min_len = lenA
            else:
                for jc in startsC:
                    target_ia = jc - lenA
                    if target_ia < 0:
                        continue
                    if lenA == 0:
                        i_start = target_ia
                    else:
                        idx = bisect.bisect_right(startsA, target_ia) - 1
                        if idx < 0:
                            continue
                        i_start = startsA[idx]
                    j = jc + lenC - 1
                    if i_start > j:
                        continue
                    length = j - i_start + 1
                    if length < min_len:
                        min_len = length

        return min_len if min_len != float('inf') else -1

# @lc code=end

File: 3455.shortest-matching-substring.cpp
#
# @lc app=leetcode id=3455 lang=cpp
#
# [3455] Shortest Matching Substring
#
# @lc code=start
class Solution {
public:
    int shortestMatchingSubstring(string s, string p) {
        int first_star = p.find('*');
        int second_star = p.find('*', first_star + 1);
        
        string prefix = p.substr(0, first_star);
        string middle = p.substr(first_star + 1, second_star - first_star - 1);
        string suffix = p.substr(second_star + 1);
        
        int n = s.length();
        
        if (prefix.empty() && middle.empty() && suffix.empty()) {
            return 0;
        }
        
        if (prefix.empty()) {
            if (middle.empty() && !suffix.empty()) {
                return s.find(suffix) != string::npos ? suffix.length() : -1;
            } else if (!middle.empty() && suffix.empty()) {
                return s.find(middle) != string::npos ? middle.length() : -1;
            } else {
                vector<int> middle_pos, suffix_pos;
                for (int i = 0; i <= n - (int)middle.length(); i++) {
                    if (s.compare(i, middle.length(), middle) == 0) {
                        middle_pos.push_back(i);
                    }
                }
                for (int i = 0; i <= n - (int)suffix.length(); i++) {
                    if (s.compare(i, suffix.length(), suffix) == 0) {
                        suffix_pos.push_back(i);
                    }
                }
                
                int min_length = INT_MAX;
                for (int m : middle_pos) {
                    for (int suf : suffix_pos) {
                        if (suf >= m + (int)middle.length()) {
                            min_length = min(min_length, suf + (int)suffix.length() - m);
                        }
                    }
                }
                return min_length == INT_MAX ? -1 : min_length;
            }
        }
        
        vector<int> prefix_pos, middle_pos, suffix_pos;
        
        for (int i = 0; i <= n - (int)prefix.length(); i++) {
            if (s.compare(i, prefix.length(), prefix) == 0) {
                prefix_pos.push_back(i);
            }
        }
        
        if (!middle.empty()) {
            for (int i = 0; i <= n - (int)middle.length(); i++) {
                if (s.compare(i, middle.length(), middle) == 0) {
                    middle_pos.push_back(i);
                }
            }
        }
        
        if (!suffix.empty()) {
            for (int i = 0; i <= n - (int)suffix.length(); i++) {
                if (s.compare(i, suffix.length(), suffix) == 0) {
                    suffix_pos.push_back(i);
                }
            }
        }
        
        int min_length = INT_MAX;
        
        for (int prefix_start : prefix_pos) {
            int min_suffix_start = prefix_start + prefix.length() + middle.length();
            
            if (suffix.empty()) {
                if (middle.empty()) {
                    min_length = min(min_length, (int)prefix.length());
                } else {
                    int min_middle_start = prefix_start + prefix.length();
                    auto it = lower_bound(middle_pos.begin(), middle_pos.end(), min_middle_start);
                    if (it != middle_pos.end()) {
                        min_length = min(min_length, *it + (int)middle.length() - prefix_start);
                    }
                }
            } else {
                for (int suffix_start : suffix_pos) {
                    if (suffix_start < min_suffix_start) continue;
                    
                    int length = suffix_start + suffix.length() - prefix_start;
                    
                    if (middle.empty()) {
                        min_length = min(min_length, length);
                    } else {
                        int min_middle_start = prefix_start + prefix.length();
                        int max_middle_start = suffix_start - middle.length();
                        
                        if (max_middle_start >= min_middle_start) {
                            auto it = lower_bound(middle_pos.begin(), middle_pos.end(), min_middle_start);
                            if (it != middle_pos.end() && *it <= max_middle_start) {
                                min_length = min(min_length, length);
                            }
                        }
                    }
                }
            }
        }
        
        return min_length == INT_MAX ? -1 : min_length;
    }
};
# @lc code=end